desktop-widgets/importgps.cpp

```diff
@@ -0,0 +1,329 @@
+int ImportGPS::get_substring(FILE *f, char *buf, char delim)
+{
+	int j;
+	char c;
+	for (j = 0; (c = fgetc(f)); j++) {
+		if (c == delim) break;
+		if (c == EOF) return 1;
+		buf[j] = c;
+		if (j >= BUFFER_SIZE)
+			return 2;
+	}
+	buf[j] = 0x0; // terminate the string
+	return 0;
+}
+
+// Find the next occurence of a specified GPX element in the file,
+// characerised by a "<xxx " or "<xxx>" character sequence.
+// s specifies the name of the element searched for.
+// termc is the ending character of the element name search = '>' or ' '.
+int ImportGPS::find_xml_element(FILE *f, const char *s, char *buf, char termc)
+{
+	bool match = false;
+	int j;
+	char c;
+	char skipdelim = (termc == ' ') ? '>' : ' ';
+	do {
+		for (j = 0; (c = fgetc(f)); j++) {  // Find the first '<' character
+			if (c == '<')
+				break;
+			if (c == EOF)
+				return 1;
+			if (j >= BUFFER_SIZE)
+				return 2;
+		}
+		for (j = 0; (c = fgetc(f)); j++) {  // Extract the xml element name after the '<'
+			if (c == EOF)
+				return 1;
+			buf[j] = c;
+			if ((c == '>') || (c == ' '))
+				break;
+		}
+		buf[j] = 0x0;	// terminate the xml element name
+		if (!(strcmp(buf, "/trk"))) // End of GPS track found: return EOF
+			return 1;
+		if (c == skipdelim) continue; // if wrong delimiter for this element was found
+		if (strcmp(s, buf))       // Compare xml element name from gpx file with the
+			match = false;    // the element searched for.
+		else
+			match = true;
+	} while (match == false);
+	return 0;
+}
+
+// Extract characters in src from first to last position 
+// e.g. substr("ascdef",dest,3,5) yields dest = "cde".
+char* ImportGPS::substr(const char *src, char *dest, int first, int last)
+{
+	char *substring = dest;
+	for (int i = first; i <= last && (*src != '\0'); i++) {
+		*substring = *(src + i);  // copy substr from src to  dest
+		substring++;
+	}
+	*substring = '\0';    // terminate and return..
+	return dest;          // ..the destination string
+}
+
+// Find the coordinates at the time specified in coords.start_dive
+// by searching the gpx file "fileName"
+int ImportGPS::getCoordsFromFile() 
+{
+	struct tm tm1;
+	time_t when = 0;
+	double lon, lat;
+	char buf[BUFFER_SIZE];
+	char strbuf[50];
+	int line=0;
+	int64_t time_offset = coords.settingsDiff_offset + coords.timeZone_offset;
+	time_t divetime;
+	bool first_line = true;
+	bool found = false;
+	QByteArray local8bitBAString1 = fileName.toLocal8Bit(); // Do these stupid gymnastics to
+	char * fname = local8bitBAString1.data();               // convert QString to a C string filename
+	divetime = coords.start_dive;
+	FILE *f1;
+	if (!(f1 = fopen(fname, "r"))) {
+		fprintf(stderr, "File open error %s\n", fname);
+		return 1;
+		}
+#ifdef GPSDEBUG
+	struct tm time; // decode the time of start of dive:
+	utc_mkdate(divetime, &time);
+	int dyr,dmon,dday,dhr,dmin;
+	dyr = time.tm_year;
+	dmon = time.tm_mon;
+	dday = time.tm_mday;
+	dhr = time.tm_hour;
+	dmin = time.tm_min;
+#endif
+
+	do {
```

Comment:
similar comment here, the buffer can be processed using Qt constructs.

Reply:
Done.